### PR TITLE
Update token name/id caching order during creation

### PR DESCRIPTION
**Pull request by MishaDemianenko**
Update order in which newly created token will be cached.
First idToToken map will be updated and then nameToId.
This will prevent cases when nameToId  already updated without corresponding token in idToToken map.
Reverse case is not possible since newly created token id is not exposed anywhere yet.

### CODE REVIEW

File: community/kernel/src/test/java/org/neo4j/kernel/TokenCreationIT.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.concurrent.locks.LockSupport;
+
+import org.neo4j.graphdb.GraphDatabaseService;
+import org.neo4j.graphdb.Label;
+import org.neo4j.graphdb.Node;
+import org.neo4j.graphdb.Transaction;
+import org.neo4j.test.EmbeddedDatabaseRule;
+import org.neo4j.test.RepeatRule;
+
+public class TokenCreationIT
```
Comment (tinwelint) [inline community/kernel/src/test/java/org/neo4j/kernel/TokenCreationIT.java]: When creating tests for very specific fixes like this I like to write a little story in the class javadoc pointing out the problem, where in the code that problem was and roughly how the test tries to expose it. It helps future maintenance.

Comment (MishaDemianenko) [inline community/kernel/src/test/java/org/neo4j/kernel/TokenCreationIT.java]: agree, story added

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/InMemoryTokenCache.java [hunk 2]
```diff
@@ -59,18 +60,18 @@ public void putAll( List<TOKEN> tokens ) throws NonUniqueTokenException
...
-        nameToId.putAll( newNameToId );
         idToToken.putAll( newIdToToken );
+        nameToId.putAll( newNameToId );
     }
 
     public void put( TOKEN token ) throws NonUniqueTokenException
     {
-        putAndEnsureUnique( nameToId, token, tokenType );
         idToToken.put( token.id(), token );
+        putAndEnsureUnique( nameToId, token, tokenType );
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/InMemoryTokenCache.java]: Why did you move the calls to `putAndEnsureUnique` after the puts to `idToToken`? If you didn't, you would not need to do the `remove` in `putAndEnsureUnique`.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/InMemoryTokenCache.java [hunk 3]
```diff
@@ -59,18 +60,18 @@ public void putAll( List<TOKEN> tokens ) throws NonUniqueTokenException
 
         for ( TOKEN token : tokens )
         {
-            putAndEnsureUnique( newNameToId, token, tokenType );
             newIdToToken.put( token.id(), token );
+            putAndEnsureUnique( newNameToId, token, tokenType );
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/InMemoryTokenCache.java]: If we get a duplicate, then this will modify the `idToToken` map by removing the existing value that conflicted with the duplicate, and then throw an exception before adding any of the new values. That's not the behaviour I would expect from a `putAll` method.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/core/InMemoryTokenCache.java [hunk 4]
```diff
@@ -43,11 +43,12 @@ public void clear()
         idToToken.clear();
     }
 
-    private static void putAndEnsureUnique( Map<String,Integer> nameToId, Token token, String tokenType )
+    private void putAndEnsureUnique( Map<String,Integer> nameToId, Token token, String tokenType )
     {
-        Integer previous;
-        if ( (previous = nameToId.put( token.name(), token.id() )) != null && previous != token.id() )
+        Integer previous = nameToId.putIfAbsent( token.name(), token.id() );
```
Comment (chrisvest) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/core/InMemoryTokenCache.java]: The `putIfAbsent` of our `CopyOnWriteHashMap` is not thread-safe.

Actually, why don't we give `CopyOnWriteHashMap` an `alter` method that is synchronised and will take a lambda of `HashMap to void`. The `HashMap` given to the lambda is then the to-be-installed copy, and the lambda can perform any modification on it that it desires. This way we can easily make arbitrary changes to a `CopyOnWriteHashMap` in a thread-safe way. Then, if we relax the generics, we can use a single map to represent both `nameToId` and `idToToken`. This way we can easily update both mappings in a thread-safe way. And the two-maps-in-one would not conflict with each other, because they use different types for their keys.

**Comment by chrisvest** [comment]
@MishaDemianenko first review done.

**Comment by MishaDemianenko** [comment]
Will just put it here. After talk with @chrisvest we decide that we will remove external synchronisation, and put both mappings into one map to make atomic updates simpler

**Comment by MishaDemianenko** [comment]
and my latest comment is not really valid, since it makes map implementation not really clear, will add additional comments to explain data flow instead.

**Comment by MishaDemianenko** [comment]
@chrisvest have added some java docs to describe current behaviour
